test_util/azure.py

```diff
@@ -0,0 +1,170 @@
+        self.ip_buckets = None
+
+    @classmethod
+    def create(
+            cls, azure_wrapper, template_uri, location, name,
+            public_key, master_prefix, agent_prefix, admin_name, oauth_enabled,
+            vm_size, agent_count, name_suffix, vm_diagnostics_enabled,
+            rollback=True):
+        assert master_prefix != agent_prefix, 'Master and agents must have unique prefixs'
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', master_prefix),\
+            'Invalid master DNS prefix: {}'.format(master_prefix)
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', agent_prefix),\
+            'Invalid agent DNS prefix: {}'.format(agent_prefix)
+        deployment_name = '{}-Deployment'.format(name)
+        group_name = '{}-ResourceGroup'.format(name)
+        if azure_wrapper.rmc.resource_groups.check_existence(group_name):
+            raise Exception("Group name already exists / taken: {}".format(group_name))
+        template = TemplateLink(uri=template_uri)
+        azure_wrapper.rmc.resource_groups.create_or_update(
+            group_name,
+            ResourceGroup(location=location))
+        template_params = {
+            'sshRSAPublicKey': public_key,
+            # must be lower or validation will fail
+            'masterEndpointDNSNamePrefix': master_prefix,
+            'agentEndpointDNSNamePrefix': agent_prefix,
+            'linuxAdminUsername': admin_name,
+            'agentVMSize': vm_size,
+            'agentCount': agent_count,
+            'nameSuffix': name_suffix,
+            'oauthEnabled': repr(oauth_enabled).lower(),
+            # oauth uses string, vm diagnostic uses bool
+            'enableVMDiagnostics': vm_diagnostics_enabled}
+        log.info('Provided template parameters: {}'.format(template_params))
+        # add worthless required 'value' fields
+        template_parameters = {k: {'value': v} for k, v in template_params.items()}
+        deployment_properties = DeploymentProperties(
+            template_link=template,
+            mode=DeploymentMode.incremental,
+            parameters=template_parameters)
+        # Use RPC against azure to validate the ARM template is well-formed
+        result = azure_wrapper.rmc.deployments.validate(
+            group_name, deployment_name, properties=deployment_properties)
+        if result.error:
+            for details in result.error.details:
+                log.error('{}: {}'.format(details.code, details.message))
+            raise Exception("Template verification failed!")
+
+        log.info('Starting template deployment')
+        deploy_poller = azure_wrapper.rmc.deployments.create_or_update(
+            group_name, deployment_name, deployment_properties)
+
+        @retry(stop_max_delay=45 * 60 * 1000,
+               retry_on_result=lambda res: res is False,
+               retry_on_exception=lambda ex: ex is not KeyboardInterrupt)
+        def poll_deploy():
```

Comment:
shouldn't polling be separate from "start it"? That way we can connect to a deployment already in process?